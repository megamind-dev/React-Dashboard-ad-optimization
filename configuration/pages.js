import {
    AUTH_FORGOT_PASSWORD_PAGE,
    AUTH_PAGE,
    AUTH_SIGNUP_PAGE,
    AUTOMATION_PAGE,
    BRAND_CAMPAIGN_KEYWORD_PAGE,
    BRAND_CAMPAIGN_PAGE,
    BRAND_CAMPAIGN_PRODUCT_AD_PAGE,
    BRAND_PAGE,
    BRANDS_SUMMARY_PAGE,
    CAMPAIGN_KEYWORD_PAGE,
    CAMPAIGN_PAGE,
    CAMPAIGN_PRODUCT_AD_PAGE,
    CAMPAIGNS_SUMMARY_PAGE,
    HOME_PAGE,
    KEYWORD_PAGE,
    KEYWORDS_SUMMARY_PAGE,
    LABELS_SUMMARY_PAGE,
    PRODUCT_PAGE,
    PRODUCT_PRODUCT_AD_PAGE,
    PRODUCTS_SUMMARY_PAGE,
    PROFILE_PAGE,
    SOV_KEYWORD_PAGE,
    SOV_KEYWORD_SEARCH_RESULT_PAGE,
    SOV_KEYWORDS_SUMMARY_PAGE,
    ORGANIZATION_PAGE,
    ORGANIZATION_GROUP_PAGE,
} from 'constants/pages'

export const PAGES_AND_PATHS = [
    {
        page: BRANDS_SUMMARY_PAGE,
        path: '/brands',
    },
    {
        page: BRAND_PAGE,
        path: '/brands/:brandId',
    },
    {
        page: BRAND_CAMPAIGN_PAGE,
        path: '/brands/:brandId/campaigns/:campaignId',
    },
    {
        page: BRAND_CAMPAIGN_KEYWORD_PAGE,
        path: '/brands/:brandId/campaigns/:campaignId/keywords/:keywordId',
    },
    {
        page: BRAND_CAMPAIGN_PRODUCT_AD_PAGE,
        path: '/brands/:brandId/campaigns/:campaignId/product-ads/:productAdId',
    },
    {
        page: CAMPAIGNS_SUMMARY_PAGE,
        path: '/campaigns',
    },
    {
        page: CAMPAIGN_PAGE,
        path: '/campaigns/:campaignId',
    },
    {
        page: CAMPAIGN_KEYWORD_PAGE,
        path: '/campaigns/:campaignId/keywords/:keywordId',
    },
    {
        page: CAMPAIGN_PRODUCT_AD_PAGE,
        path: '/campaigns/:campaignId/product-ads/:productAdId',
    },
    {
        page: PRODUCTS_SUMMARY_PAGE,
        path: '/products',
    },
    {
        page: PRODUCT_PAGE,
        path: '/products/:asin/:countryCode',
    },
    {
        page: PRODUCT_PRODUCT_AD_PAGE,
        path: '/products/:asin/:countryCode/product-ads/:productAdId',
    },
    {
        page: KEYWORDS_SUMMARY_PAGE,
        path: '/keywords',
    },
    {
        page: KEYWORD_PAGE,
        path: '/keywords/:keywordId',
    },
    {
        page: LABELS_SUMMARY_PAGE,
        path: '/labels',
    },
    {
        page: SOV_KEYWORDS_SUMMARY_PAGE,
        path: '/sov',
    },
    {
        page: SOV_KEYWORD_PAGE,
        path: '/sov/:sovKeywordId',
    },
    {
        page: SOV_KEYWORD_SEARCH_RESULT_PAGE,
        path: '/sov/:sovKeywordId/search/:scheduledDate',
    },
    {
        page: PROFILE_PAGE,
        path: '/profile',
    },
    {
        page: AUTH_PAGE,
        path: '/auth',
    },
    {
        page: AUTH_SIGNUP_PAGE,
        path: '/auth/signup',
    },
    {
        page: AUTH_FORGOT_PASSWORD_PAGE,
        path: '/auth/forgot',
    },
    {
        page: ORGANIZATION_PAGE,
        path: '/profile/organizations/:organizationId',
    },
    {
        page: ORGANIZATION_GROUP_PAGE,
        path:
            '/profile/organizations/:organizationId/groups/:organizationGroupId',
    },
    {
        page: AUTOMATION_PAGE,
        path: '/automation',
    },
    {
        page: HOME_PAGE,
        path: '/',
    },
]
