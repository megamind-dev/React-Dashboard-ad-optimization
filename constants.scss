// colors
$primary-color: #01c7d1;
$body-bg-color: #eff2f4;
$gray-background-color: #cccccc;
$white-background-color: #ffffff;
$card-border-color: #e8e8e8;
$danger-color: #f41e28;
$success-color: #52c41a;
$warning-color: #faad14;
$ghost-white: rgba(255, 255, 255, 0.2);
$text-primary: rgba(0, 0, 0, 0.85);
$text-muted: rgba(0, 0, 0, 0.65);
$text-light: rgba(0, 0, 0, 0.5);
$text-extra-light: rgba(0, 0, 0, 0.3);
$popover-mask-color: rgba(255, 255, 255, 0.5);
$side-menu-item-color: rgba(255, 255, 255, 0.5);
$action-button-color-light: #9cc681;
$action-button-color-medium: #8ab46f;
$action-button-color-dark: #68914d;

// dimensions
$row-padding: 15px;
$card-padding: 16px 24px;
$filter-panel-padding: 16px 24px;
$btn-right-margin: 10px;
$content-horiz-padding: 24px;
$icon-horizontal-margin: 5px;

// icon sizes
$medium-icon-size: 25px;
$small-icon-size: 20px;
$extra-small-icon-size: 12px;

// font sizes
$font-size-xxlarge: 28px;
$font-size-xlarge: 20px;
$font-size-large: 18px;
$font-size-medium: 14px;
$font-size-smaller: 12.6px;
$font-size-smallest: 11px;

// font weights
$font-weight-light: 300;
$font-weight-normal: 400;
$font-weight-semi-bold: 500;
$font-weight-bold: 700;

@mixin box-shadow {
  -webkit-box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16);
  -moz-box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16);
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16);
}
