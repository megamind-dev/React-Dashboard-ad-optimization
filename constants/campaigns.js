export const CAMPAIGN_BUDGET_MIN = 1
export const CAMPAIGN_BUDGET_MAX = 1000000000
export const CAMPAIGN_BUDGET_STEP = 0.1
export const CAMPAIGN_BUDGET_PRECISION = 2
