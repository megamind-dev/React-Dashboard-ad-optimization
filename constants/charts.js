/**
 * Chart Names
 */
export const ROI = 'ROI'
export const REACH = 'REACH'
export const CONVERSIONS = 'CONVERSIONS'

/**
 * Chart Axises
 */
export const LEFT_AXIS_FIRST = 'LEFT_AXIS_FIRST'
export const LEFT_AXIS_SECOND = 'LEFT_AXIS_SECOND'
export const RIGHT_AXIS = 'RIGHT_AXIS'

export const CHART_METRIC_KEYS = [LEFT_AXIS_FIRST, LEFT_AXIS_SECOND, RIGHT_AXIS]
