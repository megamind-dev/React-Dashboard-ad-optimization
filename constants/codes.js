export const REGION_CODES = {
    JP: 'JP',
    EU: 'EU',
    NA: 'NA',
}

export const REGION_LABELS = {
    [REGION_CODES.JP]: 'Japan',
    [REGION_CODES.EU]: 'Europe',
    [REGION_CODES.NA]: 'North America',
}

export const COUNTRY_CODES = {
    CA: 'CA',
    DE: 'DE',
    ES: 'ES',
    FR: 'FR',
    IN: 'IN',
    IT: 'IT',
    JP: 'JP',
    MX: 'MX',
    UK: 'UK',
    US: 'US',
}

export const COUNTRY_LABELS = {
    [COUNTRY_CODES.CA]: 'Canada',
    [COUNTRY_CODES.DE]: 'Germany',
    [COUNTRY_CODES.ES]: 'Spain',
    [COUNTRY_CODES.FR]: 'France',
    [COUNTRY_CODES.IN]: 'India',
    [COUNTRY_CODES.IT]: 'Italy',
    [COUNTRY_CODES.JP]: 'Japan',
    [COUNTRY_CODES.MX]: 'Mexico',
    [COUNTRY_CODES.UK]: 'United Kingdom',
    [COUNTRY_CODES.US]: 'United States',
}

export const LANGUAGE_CODES = {
    CES: 'CES',
    ENG: 'ENG',
    FRA: 'FRA',
    GER: 'GER',
    ITA: 'ITA',
    JPN: 'JPN',
    NLD: 'NLD',
    POL: 'POL',
    SPA: 'SPA',
    TUR: 'TUR',
    ZHO: 'ZHO',
}
