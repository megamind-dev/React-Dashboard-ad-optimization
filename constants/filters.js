export const AGGREGATION = 'aggregation'

export const BRANDS = 'brands'

export const CAMPAIGN_BUDGET = 'campaignBudget'
export const CAMPAIGN_DAYPARTINGS = 'campaignDaypartings'
export const CAMPAIGN_NAME = 'campaignName'
export const CAMPAIGN_STATES = 'campaignStates'
export const CAMPAIGN_TARGETING_TYPES = 'campaignTargetingTypes'
export const CAMPAIGNS = 'campaigns'

export const COUNTRIES = 'countries'
export const REGIONS = 'regions'

export const DATES = 'dates'

export const FACT_TYPES = 'factTypes'

export const KEYWORD_MATCH_TYPES = 'keywordMatchTypes'
export const KEYWORDS = 'keywords'

export const PRODUCT_ASINS = 'productAsins'
export const PRODUCT_TITLES = 'productTitles'

export const LABELS = 'labels'

export const SOV_AGGREGATION = 'sovAggregation'
export const SOV_COUNTRIES = 'sovCountries'
export const SOV_BRANDS = 'sovBrands'
export const SOV_FOLDS = 'sovFolds'
export const SOV_KEYWORDS = 'sovKeywords'
export const SOV_KEYWORD_CATEGORIES = 'sovKeywordCategories'
export const SOV_LANGUAGES = 'sovLanguages'
export const SOV_RESULT_TYPES = 'sovResultTypes'
export const SOV_SEARCH_TIMES = 'searchTimes'
export const SOV_STATES = 'sovStates'

export const GREATER_THAN = 'gt'
export const EQUALS = 'exact'
export const LESS_THAN = 'lt'
export const GREATER_THAN_OR_EQUAL = 'gte'
export const LESS_THAN_OR_EQUAL = 'lte'

export const NUMBER_RANGE_OPERATOR_OPTIONS = [
    { label: 'Greater Than', value: GREATER_THAN },
    { label: 'Equals', value: EQUALS },
    { label: 'Less Than', value: LESS_THAN },
    { label: 'Greater Than or Equal To', value: GREATER_THAN_OR_EQUAL },
    { label: 'Less Than or Equal To', value: LESS_THAN_OR_EQUAL },
]

export const NUMBER_RANGE_OPERATOR_DISPLAY = {
    [GREATER_THAN]: '>',
    [EQUALS]: '=',
    [LESS_THAN]: '<',
    [GREATER_THAN_OR_EQUAL]: '>=',
    [LESS_THAN_OR_EQUAL]: '<=',
}
