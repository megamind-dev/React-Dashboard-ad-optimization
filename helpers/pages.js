import { matchPath } from 'react-router-dom'
import find from 'lodash/find'
import isNull from 'lodash/isNull'
import get from 'lodash/get'
import includes from 'lodash/includes'

import { PAGES_AND_PATHS } from 'configuration/pages'
import history from 'utilities/history'
import { PROFILE_PAGE } from 'constants/pages'

export const getPath = page => get(find(PAGES_AND_PATHS, { page }), 'path')

export const getPage = (path, exact = true) => {
    let match = null
    const pageAndPath = find(PAGES_AND_PATHS, item => {
        match = matchPath(path, { path: item.path, exact })
        return !isNull(match)
    })
    if (pageAndPath) {
        const { page } = pageAndPath
        return { page, match }
    }
    return {}
}

/**
 * Get current page object
 *
 * @returns {object} - Current page object
 */
export const getCurrentPage = () => {
    const {
        location: { pathname },
    } = history

    return getPage(pathname)
}

export const isProfilePage = location => {
    const match = matchPath(location.pathname, {
        path: getPath(PROFILE_PAGE),
    })
    return Boolean(match)
}

export const isSovPage = location => {
    return includes(location.pathname, 'sov')
}
