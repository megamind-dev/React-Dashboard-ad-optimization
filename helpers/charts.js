import get from 'lodash/get'

import { AMS_TIMESERIES_CHARTS } from 'configuration/charts'

/**
 * Return default metrics of chart
 *
 * @param {string} chartName
 */
export const getChartDefaultMetrics = chartName =>
    get(AMS_TIMESERIES_CHARTS, [chartName, 'defaultMetrics'])
