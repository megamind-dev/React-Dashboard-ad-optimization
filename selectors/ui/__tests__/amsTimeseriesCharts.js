import { HOME_PAGE } from 'constants/pages'
import reducer from 'reducers'
import {
    updateChartMetrics,
    fetchHomePageHeadlineSearchTimeseriesSuccess,
    fetchHomePageSponsoredProductTimeseriesSuccess,
} from 'actions/ui'
import { LEFT_AXIS_FIRST, LEFT_AXIS_SECOND, RIGHT_AXIS } from 'constants/charts'

import {
    selectChartMetrics,
    selectDataForStackedBarChart,
} from '../amsTimeseriesCharts'
import {
    headlineSearchTimeseriesData,
    sponsoredProductTimeseriesData,
} from '../__data__/amsTimeseriesData'

describe('[Selectors] ui/amsTimeseriesCharts', () => {
    let state

    describe('selectChartMetrics', () => {
        const pageName = HOME_PAGE
        const chartName = 'timeseries'
        const metricName = 'metric'

        beforeEach(() => {
            state = reducer(
                undefined,
                updateChartMetrics({
                    pageName,
                    chartName,
                    key: metricName,
                    value: 'attributed_sales_14_day__sum',
                })
            )
        })

        it('selects chart settings by page name and chart name', () => {
            expect(selectChartMetrics(state, pageName, chartName)).toEqual({
                [metricName]: 'attributed_sales_14_day__sum',
            })
        })

        it('memoizes chart settings by page name and chart name', () => {
            const selector = selectChartMetrics.getMatchingSelector(
                state,
                pageName,
                chartName
            )

            // Reset computations
            selector.resetRecomputations()

            // Repeat computations without state change
            selectChartMetrics(state, pageName, chartName)
            selectChartMetrics(state, pageName, chartName)
            expect(selector.recomputations()).toEqual(1)

            // Update state
            state = reducer(
                state,
                updateChartMetrics({
                    pageName,
                    chartName,
                    key: metricName,
                    value: 'attributed_sales_30_day__sum',
                })
            )

            // Trigger computation after state change
            selectChartMetrics(state, pageName, chartName)
            expect(selector.recomputations()).toEqual(2)
        })
    })

    describe('selectDataForStackedBarChart', () => {
        const pageName = HOME_PAGE
        const chartName = 'timeseries'

        beforeEach(() => {
            // action to change right axis metric
            state = reducer(
                undefined,
                updateChartMetrics({
                    pageName,
                    chartName,
                    key: RIGHT_AXIS,
                    value: 'ctr',
                })
            )

            // action to change left axis first metric
            state = reducer(
                state,
                updateChartMetrics({
                    pageName,
                    chartName,
                    key: LEFT_AXIS_FIRST,
                    value: 'clicks__sum',
                })
            )

            // action to change left axis second metric
            state = reducer(
                state,
                updateChartMetrics({
                    pageName,
                    chartName,
                    key: LEFT_AXIS_SECOND,
                    value: 'impressions__sum',
                })
            )

            // action to update headline search data
            state = reducer(
                state,
                fetchHomePageHeadlineSearchTimeseriesSuccess(
                    headlineSearchTimeseriesData
                )
            )

            // action to update sponsored products data
            state = reducer(
                state,
                fetchHomePageSponsoredProductTimeseriesSuccess(
                    sponsoredProductTimeseriesData
                )
            )
        })

        it('selects chart data correctly', () => {
            const { axes, series } = selectDataForStackedBarChart(
                state,
                pageName,
                chartName
            )

            expect(axes).toHaveProperty('date')
            expect(axes).toHaveProperty('left')
            expect(axes).toHaveProperty('right')
            expect(series).toMatchSnapshot()
        })

        it('memoizes chart settings by page name and chart name', () => {
            const selector = selectDataForStackedBarChart.getMatchingSelector(
                state,
                pageName,
                chartName
            )

            // Reset computations
            selector.resetRecomputations()

            // Repeat computations without state change
            selectDataForStackedBarChart(state, pageName, chartName)
            selectDataForStackedBarChart(state, pageName, chartName)
            expect(selector.recomputations()).toEqual(1)

            // Update state
            state = reducer(
                state,
                updateChartMetrics({
                    pageName,
                    chartName,
                    key: RIGHT_AXIS,
                    value: 'ctr',
                })
            )

            // Trigger computations after state change
            selectDataForStackedBarChart(state, pageName, chartName)
            selectDataForStackedBarChart(state, pageName, chartName)
            expect(selector.recomputations()).toEqual(2)
        })
    })
})
