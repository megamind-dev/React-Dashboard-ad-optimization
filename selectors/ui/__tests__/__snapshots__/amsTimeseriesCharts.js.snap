// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Selectors] ui/amsTimeseriesCharts selectDataForStackedBarChart selects chart data correctly 1`] = `
Object {
  "headlineSearch": Object {
    "LEFT_AXIS_FIRST": Array [
      Array [
        1542672000000,
        2480,
      ],
      Array [
        1542758400000,
        2420,
      ],
      Array [
        1542844800000,
        2568,
      ],
      Array [
        1542931200000,
        3972,
      ],
    ],
    "LEFT_AXIS_SECOND": Array [
      Array [
        1542672000000,
        221219,
      ],
      Array [
        1542758400000,
        218753,
      ],
      Array [
        1542844800000,
        237866,
      ],
      Array [
        1542931200000,
        357965,
      ],
    ],
    "RIGHT_AXIS": Array [
      Array [
        1542672000000,
        0.011210610300200255,
      ],
      Array [
        1542758400000,
        0.011062705425754162,
      ],
      Array [
        1542844800000,
        0.01079599438339233,
      ],
      Array [
        1542931200000,
        0.011096056877068987,
      ],
    ],
  },
  "sponsoredProducts": Object {
    "LEFT_AXIS_FIRST": Array [
      Array [
        1542672000000,
        12018,
      ],
      Array [
        1542758400000,
        11823,
      ],
      Array [
        1542844800000,
        14030,
      ],
      Array [
        1542931200000,
        21651,
      ],
    ],
    "LEFT_AXIS_SECOND": Array [
      Array [
        1542672000000,
        2397061,
      ],
      Array [
        1542758400000,
        2362859,
      ],
      Array [
        1542844800000,
        3247452,
      ],
      Array [
        1542931200000,
        5967377,
      ],
    ],
    "RIGHT_AXIS": Array [
      Array [
        1542672000000,
        0.0050136396195174,
      ],
      Array [
        1542758400000,
        0.005003684096257966,
      ],
      Array [
        1542844800000,
        0.004320310200119971,
      ],
      Array [
        1542931200000,
        0.0036282272764063673,
      ],
    ],
  },
}
`;
